arch: arm: Initialize floating point registers for every CPU with a FPU

PierreAronnax opened the pull request:
This restores the registers back to reset value, even if CONFIG_FLOAT
is not set.
These registers my be changed earlier for example by a bootloader. This
caused stack offset when booting from a build-in EFM32GG bootloader.

Clearing CPACR unconditionally supports switching between Full access
and Privileged access only.

Initialize the Floating Point Status and Control Register before the
Instruction Synchronization Barrier for symmetry with
arch_switch_to_main_thread.

Should support all combinations of CONFIG_FLOAT and CONFIG_FP_SHARING

Fixes #22977

Signed-off-by: Luuk Bosma <l.bosma@interay.com>

ioannisg reviewed (changes requested):
Thanks for the patch:
I think there are two different bugs here
- CPACR is not cleared before it is ORed with CP10/CP11 Full/PRIV access
- CONTROL is not clearing the FPCA bit at boot

I am OK with how you address the first bug, i.e. clearing CPACR before it is ORed. Can you update also the sentence that "Upon reset, the Co-Processor Access Control Register is 0x00000000" and add a nore explaining that "we do not rely on reset to clear the register? Then it will be more intuitive _why_ the clearing is done. Can you do that on a separate commit?

Then, regarding the second bug: 

Though your solution may work, I see that we already do the CONTROL FPCA bit clearing when switching to the main thread (under FP_SHARING):

```
void arch_switch_to_main_thread(struct k_thread *main_thread,
				k_thread_stack_t *main_stack,
				size_t main_stack_size,
				k_thread_entry_t _main)
{
#if defined(CONFIG_FLOAT)
	/* Initialize the Floating Point Status and Control Register when in
	 * Unshared FP Registers mode (In Shared FP Registers mode, FPSCR is
	 * initialized at thread creation for threads that make use of the FP).
	 */
	__set_FPSCR(0);
#if defined(CONFIG_FP_SHARING)
	/* In Sharing mode clearing FPSCR may set the CONTROL.FPCA flag. */
	__set_CONTROL(__get_CONTROL() & (~(CONTROL_FPCA_Msk)));
	__ISB();
#endif /* CONFIG_FP_SHARING */
#endif /* CONFIG_FLOAT */

```

Perhaps you could re-organize the #ifdef's there, so the FPCA is cleared un-conditionally, when building with CPU_HAS_FPU.

Alternatively, you could clear the CONTROL.FPCA in prepC, but I would like this to be done outside the enable_floating_point() function - as this is not-specific to when building with FLOAT support. This solution will, however, make the FPCA bit cleared two times

ioannisg committed:
tests: arch: arm_thread_swap: test FPSCR preservation in FP_SHARING

Under Sharing FP registers mode we would like to verify the
correct preservation of FPSCR during thread context switch.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

PierreAronnax commented:
I am working out your solution.

But wouldn't it be more logical to keep the initialization in enable_floating_point (with updated comments) and completely remove it from arch_switch_to_main_thread instead?

Your solution splits the initialization over two functions, but for example Initialize the Floating Point Status and Control Register `__set_FPSCR` was already cleared two times.

PierreAronnax commented:
**Solution 1**
https://github.com/zephyrproject-rtos/zephyr/pull/23008/commits/aaf7919fd0dced09618319e229c2ea25ca9d4ed4
Updated comments and removing initialization from arch_switch_to_main_thread to avoid initialization two times.

**Solution 2**
https://github.com/PierreAronnax/zephyr/commit/59075ee030a94e475a255ef5a1494093ae42c892
Updated according your proposal.

Still not sure about the comments…
Your thoughts?

ioannisg commented:
> I am working out your solution.
> 
> But wouldn't it be more logical to keep the initialization in enable_floating_point (with updated comments) and completely remove it from arch_switch_to_main_thread instead?
> 
> Your solution splits the initialization over two functions, but for example Initialize the Floating Point Status and Control Register `__set_FPSCR` was already cleared two times.

There is a reason for doing that, pls, check the relevant commit messages :)
We want to clear FPSCR at boot, in case floating point is used during the initialization of zephyr (e.g. drivers). So we need to clear it again before main.

ioannisg commented:
@PierreAronnax I'll review your Solutions, 1, and 2.
You can see how I was thinking of fixing the bug here: https://github.com/zephyrproject-rtos/zephyr/pull/23074

PierreAronnax commented:
I have verified your PR and it looks good to me.
However, your solution does not disable the CP10 and CP11 Co-Processors when CONFIG_FLOAT=n
Upon boot from bootloader Co-Processor Access Control Register is 0xf00000

ioannisg commented:
> I have verified your PR and it looks good to me.
> However, your solution does not disable the CP10 and CP11 Co-Processors when CONFIG_FLOAT=n
> Upon boot from bootloader Co-Processor Access Control Register is 0xf00000

Agreed. But, if CONFIG_FLOAT=n, the FP extension is not used, anyways.

ioannisg commented:
> I have verified your PR and it looks good to me.
Would you like to adopt the solution approach in #23074 in your PR then?

PierreAronnax commented:
Please take over this PR and use your PR instead.

ioannisg commented:
> Please take over this PR and use your PR instead.

Oh I was thinking you'd like to be the author of this bug-fix.

PierreAronnax commented:
Alright, will do tomorrow.

ioannisg commented:
> Alright, will do tomorrow.

OK great @PierreAronnax , we 'll wait for this bug-fix :)
Adding v2.2 milestone, to get this intor v2.2.0 release

LuukBosma committed:
arch: arm: aarch32: clear CPACR for every CPU that has a FPU

Upon reset, the Co-Processor Access Control Register is, normally,
0x00000000. However, it might be left un-cleared by firmware running
before Zephyr boot.
This restores the register back to reset value, even if CONFIG_FLOAT
is not set.
Clearing before setting supports switching between Full access
and Privileged access only.

Refactor enable_floating_point to support initialize
floating point registers for every CPU that has a FPU.

Signed-off-by: Luuk Bosma <l.bosma@interay.com>

LuukBosma committed:
arch: arm: aarch32: clear CONTROL.FPCA for every CPU that has a FPU

Upon reset, the CONTROL.FPCA bit is, normally, cleared. However,
it might be left un-cleared by firmware running before Zephyr boot,
for example when Zephyr image is loaded by another image.
We must clear this bit to prevent errors in exception unstacking.
This caused stack offset when booting from a build-in EFM32GG bootloader

Fixes #22977

Signed-off-by: Luuk Bosma <l.bosma@interay.com>

PierreAronnax commented:
I have made the changes, limited line length etc.
I also included your test, was that the intention?

ioannisg commented:
> I have made the changes, limited line length etc.
> I also included your test, was that the intention?

Yes, it is good to have these checks in regression. 
Do you have a platform to run this test and verify it is passing?

PierreAronnax commented:
Board: efm32gg_stk3701a
CONFIG_FLOAT=y
CONFIG_FP_SHARING=y
```
*** Booting Zephyr OS build v2.2.0-rc1-262-g0d7e3dcc79ef  ***
Running test suite arm_thread_swap
===================================================================
starting test - test_arm_thread_swap
PASS - test_arm_thread_swap
===================================================================
Test suite arm_thread_swap succeeded
===================================================================
PROJECT EXECUTION SUCCESSFUL
```

ioannisg reviewed (approved):
Thanks for the bug report and the fix PR
